src/lib/reasoners/bitv_rel.ml

```diff
@@ -428,55 +643,256 @@ module Any_constraint = struct
+        Domains.Ephemeral.structural_propagation
+        (QC.pop queue) dom
+    done
+  with QC.Empty -> ()
+
+let propagate_intervals queue touched bcs dom =
+  let touch_c c = QC.push queue (Constraint c) in
+  let touch r =
+    HX.replace touched r ();
+    QC.push queue (Structural r);
+    Constraints.iter_parents touch_c r bcs
+  in
+  try
+    while true do
+      Interval_domains.Ephemeral.iter_changed touch dom;
+      Interval_domains.Ephemeral.clear_changed dom;
+      Any_constraint.propagate
+        Constraint.propagate_interval
+        Interval_domains.Ephemeral.structural_propagation
+        (QC.pop queue) dom
+    done
+  with QC.Empty -> ()
+
+let propagate_all eqs bcs bdom idom =
   (* Call [simplify_pending] first because it can remove constraints from the
      pending set. *)
   let eqs, bcs = Constraints.simplify_pending eqs bcs in
   (* Optimization to avoid unnecessary allocations *)
-  if Constraints.has_pending bcs || Domains.has_changed dom then
+  let do_all = Constraints.has_pending bcs in
+  let do_bitlist = do_all || Domains.has_changed bdom in
+  let do_intervals = do_all || Interval_domains.has_changed idom in
+  let do_any = do_bitlist || do_intervals in
+  if do_any then
     let queue = QC.create 17 in
-    let touch_c c = QC.push queue (Constraint c) in
-    Constraints.iter_pending touch_c bcs;
+    let touch_pending queue =
+      Constraints.iter_pending (fun c -> QC.push queue (Constraint c)) bcs
+    in
+    let bitlist_changed = HX.create 17 in
+    let touched = HX.create 17 in
+    let bdom = Domains.edit bdom in
+    let idom = Interval_domains.edit idom in
+
+    (* First, we propagate the pending constraints to both domains. Changes in
+       the bitlist domain are used to shrink the interval domains. *)
+    touch_pending queue;
+    propagate_bitlist queue touched bcs bdom;
+    assert (QC.is_empty queue);
+
+    touch_pending queue;
+    HX.iter (fun r () ->
+        HX.replace bitlist_changed r ();
+        constrain_interval_from_bitlist
+          Interval_domains.Ephemeral.(handle idom r)
+          Domains.Ephemeral.(!!(handle bdom r))
+      ) touched;
+    HX.clear touched;
+    propagate_intervals queue touched bcs idom;
+    assert (QC.is_empty queue);
+
+    (* Now the interval domain is stable, but the new intervals may have an
+       impact on the bitlist domains, so we must shrink them again when
+       applicable. We repeat until a fixpoint is reached. *)
     let bcs = Constraints.clear_pending bcs in
-    let changed = HX.create 17 in
-    let touch r =
-      HX.replace changed r ();
-      QC.push queue (Structural r);
-      Constraints.iter_parents touch_c r bcs
+    while HX.length touched > 0 do
+      HX.iter (fun r () ->
+          constrain_bitlist_from_interval
+            Domains.Ephemeral.(handle bdom r)
+            Interval_domains.Ephemeral.(!!(handle idom r))
+        ) touched;
+      HX.clear touched;
+      propagate_bitlist queue touched bcs bdom;
```

Comment:
I don't understand why the queue isn't necessary empty here? We have flushed `bcs` before this loop and `bcs` is immutable and the queue is empty before the loop too.